### ISSUE TITLE
Option to turn off logging for invalid IP message

### ISSUE THREAD

**Issue by jhitze** [logstash-plugins/logstash-filter-geoip#29]
What do you all think of having the "IP Field contained invalid IP address or hostname" message be able to turn off?

Like, I know it's going to happen in my IIS logs, as it will put a `-` field in there if the event doesn't have an IP. 

The reason I ask, is because my logstash logs are at 25gb in just a week, mostly because of this log entry.

```shell
grep "IP Field contained invalid IP address or hostname" * | wc -l
375898
```
And that count doesn't even count the number of log lines associated with each of those entries! :(

I'm thinking that it should default to true, so that anyone trying it out for the first time would get the error message like today. Then allow it to be turned off in the config.

```ruby
geoip {
  source => "ip"
  invalid_ip_logging => false
}
```

**Comment by splashx** [comment]
This error also happens if the logstash instance is running in an environment where there is no possibility to resolve the external hostnames.

**Comment by ricardojdsilva87** [comment]
Hello,

I'm having the same problem, but my server has connection to the internet so I can resolve the hostnames on the apache logs.

I have some IP/hostnames on the apache logs that look like this:
115031194-171.lexues.net
213-195-147-170.static.ip.netia.com.pl´
ul.cs-cart.com

When passing trough the geoip filter on the logstash.log I can see the error "IP Field contained invalid IP address or hostname". 

Checking on the machine I can't resolve the hostname 115031194-171.lexues.net into an IP address and to a location, but using only lexues.net the call is valid and I get an IP address. The same applies to the other examples.

My question is if there is any way of trying to resolve those names, or at least using the domain name instead. All this without having to try to identify the correct dns name from a long string, for example lexues.net.

Thanks for the help

**Comment by ipbastola** [comment]
I am facing the similar issue. Any help?

**Comment by ebuildy** [comment]
As you can see on sources:

https://github.com/logstash-plugins/logstash-filter-geoip/blob/91d35650f6defe80b9d1f9a53edb34db6868dcd2/lib/logstash/filters/geoip.rb#L160

It's using logstash logger, 2 solutions:
- Comment this line ^^
- Disable logstash logging

**Comment by eoconnor** [comment]
Another option would be to only call the geoip filter if the field value matches an IP regular expression:

`
if [my-ip-field] =~ "[0-9]*.[0-9]*.[0-9]*.[0-9]*" {
    geoip {
        source => "my-ip-field"
    }
}
`

**Comment by jordansissel** [comment]
This feels a bit odd. The intent of this plugin is to act on IP addresses, so it's up to the user to only send IP addresses. We could change this behavior, though none of the proposed solutions (comment it out, add a new setting) feel like the right solution.

We could move this log from `error` level to `info` level, which would make it not logged by default. THoughts?

**Comment by eoconnor** [comment]
> The intent of this plugin is to act on IP addresses, so it's up to the user to only send IP addresses.

This works great if you have direct control over the data flowing into your Logstash cluster. I don't, and I suspect others don't as well.

That being said, a new config param that allows you to set the logging level of this message could be appropriate. In our case, we don't particularly care if we get the occasionally bad value in the field we're passing to the filter, so the if statement approach works just fine for us.

**Comment by TinLe** [comment]
It would be nice to be able to tell the filter to ignore/drop invalid IP addresses.   I ended up putting this mutate filter right before my geoip {}

>             mutate {
>                 gsub => [
>                     "client_ip", "-", "127.0.0.1"
>                 ]
>             }

Tin

**Comment by pemontto** [comment]
:+1:

**Comment by omercnet** [comment]
+1

**Comment by dellis23** [comment]
In my case, it's less about invalid IPs and more about them being IPv6, which doesn't seem to be supported.  So it'd be nice to have a way to disable logging of this error.

**Comment by rahulsinghai** [comment]
👍

**Comment by dacofr** [comment]
:+1: to add an option to disable log.

I've the same problem when they are multiple IP in the field :

```
"IP"=>"192.168.10.15, 81.248.104.152"
```

**Comment by aramirez-es** [comment]
👍  to disable log.

I'm using GeoIp in `X-Forwarded-For` field because my servers are placed behind Amazon Cloud Front and ELB. As you know, what `X-Forwarded-For` header may contain does not guarantee to be a right IP address: you could get '-', 'unknown' or any kind of bad formatted string.

So currently, the two available options are:

1. Using mutate filter to replace strings by the local IP.
2. Check if the field is a correct IP (using regex, what it seems to me even worse) before applying GeoIP plugin.

So, IMO the best option would be to disable log. Since we can **tag** log lines if geoip filter is successful, we don't miss errors.

**Comment by ebuildy** [comment]
I think this is not geoip-filter problem, you should check before the IP is valide or not.

You could use a filter with inline ruby to check if IP is valid or not, explode field etc... but geoip is not the right place to do it.

**Comment by ebuildy** [comment]
Of course yes, in any system, if you enter wrong value you got an error! If you don't want the error, you must validate your data before, this is quite common in web dev.

This is "Single responsibility principle" pattern.

2 lines of logstash to get rid of this, quite easy, you wrote more on github ;-) 

Or feel free to send a PR.

**Comment by aramirez-es** [comment]
@ebuildy you are right, any system must validate incoming data, I'm not saying otherwise. But, as client of a library/package/system, I want ability to disable logs or catch errors (like you can do either in any programming language or using any logging framework: you as client decide the logging/error display policy).

Take a look to the following example:

```
grok {
    match => { "message" => '%{COMBINEDAPACHELOG} %{DATA:xforwardedfor}' }
}

split => ["xforwardedfor", ","]
strip => ["xforwardedfor"]
gsub => [ "xforwardedfor", "-|unknown", "127.0.0.1" ]

geoip {
    source => "xforwardedfor"
}
```

Of course I could use `IP` instead of `DATA` but if I do so, Grok wouldn't match things like `IP1, IP2`, which are quite common.

In the other hand, the `gsub` filter is not enough because I found `X-Forwarded-For` headers like:
- u08022, 146.152.212.37, 36.119.253.28
- u60521, 12.20.89.121, 188.13.109.176
- 55.33.131.161:47168, 62.254.110.44

*notice the port in the last example*

Instead of repeating the IP validation in every client, I think turning off the logging for those who don't care failures, it would be the best choice IMO.

Anyway, thank you for your time :)

**Comment by ebuildy** [comment]
You mean something like https://github.com/logstash-plugins/logstash-filter-geoip/pull/97/files ?

**Comment by rparkhunovsky** [comment]
Any chance for merging https://github.com/logstash-plugins/logstash-filter-geoip/pull/97?
